### PR TITLE
Fix #55, issue with sourcemap output

### LINKED ISSUE #55: source maps are incorrect when compiling from SCSS
Sorry--this is actually the same as issue 54 but I'm not sure how to reopen it. As mentioned there, the problem does not lie with the other plugins in the compile chain. The issue is with this plugin because it relies on `gulp-postcss`, which is the ultimate culprit. The issue is easily reproduced with `gulp-autoprefixer` and any number of other plugins.

I have this Gulp code which compiles three SCSS files (first.scss, last.scss, and app.scss). All plugins are gulp-sourcemaps compatible:

``` javascript
return gulp.src(cssFiles)
    .pipe(plugins.if(addSourceMaps, plugins.sourcemaps.init({loadMaps: true})))
    .pipe(plugins.sass())
    .pipe(plugins.autoprefixer())

    // I have verified this code is not the issue--if commented out the problem remains.
    .pipe(plugins.if(!!config.cssFile,plugins.order(
      [
        config.sourceFolder + '**/first.css',
        '!' + config.sourceFolder + '**/*',
        '!' + config.sourceFolder + '**/last.css'
      ],
      {base: '.'}
    )))

    .pipe(plugins.if(!!config.cssFile, plugins.concat(config.cssFile)))
    .pipe(plugins.if(addSourceMaps, plugins.sourcemaps.write('.', {sourceRoot: '../src'})))
    ...
```

When autoprefixer is commented out, the source maps are produced properly:

```
{"version":3,"sources":["sass/first.scss","sass/app.scss","sass/last.scss"],"names":[],"mappings":"AAAA;ACAA;ACAA","file":"all.css","sourcesContent":["// All CSS files are combined into one big file called \"all.css\". This file is\n// guaranteed to be the first file added (the contents of this file will be at\n// the beginning of \"all.css\").\n\n","// app.scss\n\n","// All CSS files are combined into one big file called \"all.css\". This file is\n// guaranteed to be the last file added (the contents of this file will be at\n// the end of \"all.css\").\n\n"],"sourceRoot":"../src"}
```

When autoprefixer is enabled, the source maps are incorrect and point to nonexistent CSS files:

```
{"version":3,"sources":["sass/app.css","sass/first.css","sass/last.css"],"names":[],"mappings":"AAAA;ACAA;ACAA","file":"all.css","sourceRoot":"../src","sourcesContent":[]}
```

### PR DESCRIPTION
**Pull request by Saturate**
This fixes issue #55 for me with css generated from less. Code adapted from `gulp-postcss`

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -22,10 +23,16 @@ module.exports = function (opts) {
...
+			var map;
 			file.contents = new Buffer(res.css);
 
 			if (res.map && file.sourceMap) {
-				applySourceMap(file, res.map.toString());
+				map = res.map.toJSON();
+				map.file = file.relative;
+				map.sources = [].map.call(map.sources, function (source) {
+					return path.join(path.dirname(file.relative), source);
+				});
```
Comment (sindresorhus) [inline index.js]: Why is this needed? We already pass the `file.path` to the postcss function. Seems rather like an issue with either `postcss` or `vinyl-sourcemaps-apply`.

Comment (Saturate) [inline index.js]: It works without, but the reason `gulp-postcss` is doing it is that it's missing the path. If we don't do this, all files will look like they are coming from the root of the website.

I did it mostly due to `gulp-postcss` doing the same thing, and they are maintaining both `postcss` and `gulp-postcss` seeming that they think it's the right solution. But one can argue that the problem is postcss.

I can remove it if you like :-)

Comment (Saturate) [inline index.js]: They added it in this commit: https://github.com/postcss/gulp-postcss/commit/9feb6d4f99b705a4e9070abbd7dfa06a2d31b3b2#diff-168726dbe96b3ce427e7fedce31bb0bc to fix the relative sourcemaps.

Comment (sindresorhus) [inline index.js]: No, I mean the whole change. Why is any of this needed. It should just work. I don't think it makes sense to work around stuff here. It should rather be fixed upstream so it benefits all consumers.

Comment (sindresorhus) [inline index.js]: If it's a requirement to make the sourcemap paths relative, I think it's the job of https://github.com/floridoo/vinyl-sourcemaps-apply which already get's the `file` object and should handle any sourcemap related task.

Comment (Saturate) [inline index.js]: Sorry for being a bit unclear. Without these changes it's simply not working right now.

Right now it only keeps the autoprefixer sourcemaps, and throws away the LESS ones, it outputs this:

```
version: 3,
sources: [
   "top-navigation.css"
],
names: [ ],
mappings: "..",
file: "/Styles/Domain/Navigation/top-navigation.css",
sourcesContent: [
   "..."
],
sourceRoot: "/source/"
}
```

Giving this in DevTools, that's not useful at all since it's not the real source.
image

After this patch it gives this:

```
version: 3,
sources: [
   "/Styles/Domain/Navigation/top-navigation.css",
   "/Styles/Domain/Navigation/top-navigation.less",
   "/Styles/TextMixins.less",
   "/Styles/BoxMixins.less"
],
names: [ ],
mappings: "..",
file: "/Styles/Domain/Navigation/top-navigation.css",
sourcesContent: [
   "...",
   "...",
   "...",
   "..."
],
sourceRoot: "/source/"
}
```
that's translated to the real source in DevTools:

image

Hope this makes a bit more sense.

Comment (sindresorhus) [inline index.js]: I agree it's an issue. I'm just saying you're trying to fix it in the wrong place.

Comment (Saturate) [inline index.js]: I linked the issue and this PR on the postcss issue (https://github.com/postcss/postcss/issues/240), the information within this PR might be useful.
